lib/Dialect/SPIRV/SPIRVOps.cpp

```diff
@@ -544,6 +544,62 @@ static LogicalResult verify(spirv::AccessChainOp accessChainOp) {
   return success();
 }
 
+namespace {
+
+// Combine chained `spirv::AccessChainOp` operations into one
+// `spirv::AccessChainOp` operation. The current approach is based on analyzing
+// only two operations at once: the current `spirv::AccessChainOp` and an
+// operation immediately following the current operation.
+struct AccessChainOpCanonicalizer
+    : public OpRewritePattern<spirv::AccessChainOp> {
+  using OpRewritePattern<spirv::AccessChainOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(spirv::AccessChainOp accessChainOp,
+                                     PatternRewriter &rewriter) const override {
+    // It is safe to look at the next operation from the current insertion
+    // point, because each block must have a terminator operation.
+    auto nextAccessChainOp = dyn_cast<spirv::AccessChainOp>(
+        *std::next(rewriter.getInsertionPoint()));
+
+    if (!nextAccessChainOp ||
+        !canCanonicalizeChainedAccesses(accessChainOp, nextAccessChainOp)) {
+      return matchFailure();
+    }
+    // Combine indices.
+    llvm::SmallVector<Value *, 4> indices(accessChainOp.indices());
+    indices.insert(indices.end(), nextAccessChainOp.indices().begin(),
+                   nextAccessChainOp.indices().end());
+
+    // Replace `nextAccessChainOp` with newly created `spirv::AccessChainOp`.
+    rewriter.replaceOpWithNewOp<spirv::AccessChainOp>(
```

Comment:
Hmm, normally an op's canonicalization should only change the op itself. So instead of looking at and replacing the **next** access chain op, what about just checking whether the current access chain op's base pointer to see whether it comes from an access chain op, if so, replace the **current** access chain op by using the previous access chain op's base pointer and all the indices and leave the previous access chain op untouched? That is, turning

```
%0 = spv.AccessChain %ptr, %idx0, %idx1
%1 = spv.AccessChain %0, %idx2, %idx3
```
to
```
%0 = spv.AccessChain %ptr, %idx0, %idx1
%1 = spv.AccessChain %ptr, %idx0, %idx1, idx2, %idx3
```

The previous access chain op may have multiple users and that's fine. If it has no additional users then the framework will handle the removal of it directly.

Reply:
@antiagainst thanks for point me on this!